From 11bd0d18cb6cf2f7ad3092330419022b8ed169e5 Mon Sep 17 00:00:00 2001
From: Stephen Chen <stephen@radxa.com>
Date: Sat, 28 Jan 2023 19:09:25 +0800
Subject: [PATCH 304/364] ethernet: stmmac: add phy rtl8211f register fixup

RTL8211F LED indication
  * LED1(Green) LED : Link + Active
  * LED2(Yellow) LED: Link

Signed-off-by: Stephen Chen <stephen@radxa.com>
(cherry picked from commit 9ffaf37a55345eb3c2f70aca00eaa0d88876c53e)
Signed-off-by: Joshua Riek <jjriek@verizon.net>
---
 .../net/ethernet/stmicro/stmmac/stmmac_main.c | 39 +++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
index a6310d96b0d4..8ebfd61f32ca 100644
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -60,6 +60,12 @@
 #define	STMMAC_ALIGN(x)		ALIGN(ALIGN(x, SMP_CACHE_BYTES), 16)
 #define	TSO_MAX_BUFF_SIZE	(SZ_16K - 1)
 
+#define	RTL8211F_PHY_ID		0x001cc916
+#define	RTL8211F_PHY_ID_MASK	0x001fffff
+#define	RTL8211F_PAGE_SELECT	0x1f
+#define	RTL8211F_LCR_ADDR	0x10
+#define	RTL8211F_EEELCR_ADDR	0x11
+
 /* Module parameters */
 #define TX_TIMEO	5000
 static int watchdog = TX_TIMEO;
@@ -7097,6 +7103,33 @@ void stmmac_fpe_handshake(struct stmmac_priv *priv, bool enable)
 	}
 }
 
+static int phy_rtl8211f_led_fixup(struct phy_device *phydev)
+{
+	u32 val, val2;
+
+	/* Switch to Page 0x0d04 */
+	phy_write(phydev, RTL8211F_PAGE_SELECT, 0x0d04);
+
+	/* Set LED1(Green) Link 10/100/1000M + Active, and set LED2(Yellow) Link 10/100/1000M */
+	val = phy_read(phydev, RTL8211F_LCR_ADDR);
+	val |= (1<<5);
+	val |= (1<<8);
+	val |= (1<<10);
+	val |= (1<<11);
+	val &= (~(1<<14));
+	phy_write(phydev, RTL8211F_LCR_ADDR, val);
+
+	/* Disable LED2(Yellow) EEE LED function to keep it on when linked */
+	val2 = phy_read(phydev, RTL8211F_EEELCR_ADDR);
+	val2 &= (~(1<<3));
+	phy_write(phydev, RTL8211F_EEELCR_ADDR, val2);
+
+	/* Switch back to the PHY's IEEE Standard Registers. Here it is Page 0 */
+	phy_write(phydev, RTL8211F_PAGE_SELECT, 0);
+
+	return 0;
+}
+
 /**
  * stmmac_dvr_probe
  * @device: device pointer
@@ -7378,6 +7411,12 @@ int stmmac_dvr_probe(struct device *device,
 	 */
 	pm_runtime_put(device);
 
+	/* Register fixup for PHY RTL8211F */
+	ret = phy_register_fixup_for_uid(RTL8211F_PHY_ID, RTL8211F_PHY_ID_MASK, phy_rtl8211f_led_fixup);
+	if (ret) {
+		dev_warn(priv->device, "Failed to register fixup for PHY RTL8211F.\n");
+	}
+
 	return ret;
 
 error_netdev_register:
-- 
2.43.0

